groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+union VariantUnion {
+};
+
+template <class HEAD, class... TAIL>
+union VariantUnion<HEAD, TAIL...> {
+    typedef bsl::allocator<char> allocator_type;
+    Variant_DataImp<HEAD>        d_head;
+    VariantUnion<TAIL...>        d_tail;
+
+    VariantUnion(){};
+    // Construct a VariantUnion without an active alternative type.
+
+    template <class... ARGS>
+    VariantUnion(bsl::in_place_index_t<0>,
+                 BSLS_COMPILERFEATURES_FORWARD_REF(ARGS)... args)
+    : d_head(BSLS_COMPILERFEATURES_FORWARD(ARGS, args)...)
+    {
+        // Construct a VariantUnion with the current alternative as the active
+        // one.  Note that the 'in_place_index' does not specify the absolute
+        // index of the alternative.  We use the tag to 'unravel' the union
+        // until we get to the correct alternative type.
+    }
+
+    template <size_t N, class... ARGS>
+    VariantUnion(bsl::in_place_index_t<N>,
+                 BSLS_COMPILERFEATURES_FORWARD_REF(ARGS)... args)
+    : d_tail(bsl::in_place_index_t<N - 1>(),
+             BSLS_COMPILERFEATURES_FORWARD(ARGS, args)...)
+    {
+        // Continue unraveling the VariantUnion until you reach the alternative
+        // corresponding to the given index.
+    }
+};
+#else  //BSL_VARIANT_FULL_IMPLEMENTATION
+
+template <class TYPE>
+class Variant_DataImp {
+    // This component-private 'struct' manages a 'value_type' object in an
+    // 'variant' object.  This class provides an abstraction for 'const' value
+    // type.  A 'variant' object may contain an object of 'const' type.  An
+    // assignment to such an 'variant' object should not succeed.  However,
+    // unless the 'variant' object itself is 'const', it should be possible to
+    // change the value of the 'variant' object using 'emplace'.  In order to
+    // allow for that, this class manages a non-const object of 'value_type',
+    // but all the accessors return a 'const' adjusted reference to the managed
+    // object.
+
+  private:
+    // PRIVATE TYPES
+    typedef typename bsl::remove_const<TYPE>::type StoredType;
+    typedef typename bsl::allocator<char>          allocator_type;
+    typedef BloombergLP::bslmf::MovableRefUtil     MoveUtil;
+
+    // DATA
+    bsls::ObjectBuffer<StoredType> d_buffer;
+        // in-place 'TYPE' object
+
+  public:
+    // CREATORS
+
+    // MANIPULATORS
+    TYPE& value();
+        // Return the 'value_type' object in 'd_buffer' with const
+        // qualification adjusted to match that of 'TYPE'.  The behavior is
+        // undefined unless the object holds the currently active alternative.
+
+    const TYPE& value() const;
+        // Return the 'value_type' object in 'd_buffer' with const
+        // qualification adjusted to match that of 'TYPE'.  The behavior is
+        // undefined unless the object holds the currently active alternative.
+};
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+template <class HEAD = void, class... ARGS>
+union VariantUnion;
```

Comment:
Bummer. Can you please add a comment to that effect, in order to prevent the next person who looks at this from trying to "simplify" it? Something like "Note that the declaration and definition of the primary template must be kept separate in order for sim_cpp11_features.pl to work properly."

Reply:
done